### PR TITLE
GH-38438: [C++] Dataset: Trying to fix the async bug in Parquet dataset

### LINKED ISSUE #38438: [C++][Python] Segfault during `pyarrow.dataset.write_dataset` with dataset source read with pre_buffer=True
### Describe the bug, including details regarding any error messages, version, and platform.

Please see https://github.com/voltrondata-labs/arrow-benchmarks-ci/issues/166 for further detail.

Since https://github.com/apache/arrow/pull/37854, the arrow-benchmarks-ci runners have been attempting to run the `dataset-serialize` [benchmark](https://github.com/voltrondata-labs/benchmarks/blob/93a57ba1019f65a4cea124d17772b0e8c60ac5ce/benchmarks/dataset_serialize_benchmark.py) on an x86_64 Ubuntu runner using Python 3.8. Each time, somewhere between 0 and 3 cases succeed before we see `Fatal Python error: Segmentation fault`.

### Component(s)

Benchmarking, Python

### PR DESCRIPTION
**Pull request by mapleFU**
### Rationale for this change

Origin mentioned https://github.com/apache/arrow/issues/38438

1. When PreBuffer is default enabled, the code in `RowGroupGenerator::FetchNext` would switch to async mode. This make the state handling more complex
2. In `RowGroupGenerator::FetchNext`, `[this]` is captured without `shared_from_this`. This is not bad, however, `this->executor_` may point to a invalid address if this dtor.

This patch also fixes a lifetime issue I founded in CSV handling.

### What changes are included in this PR?

1. Fix handling in `cpp/src/parquet/arrow/reader.cc` as I talked above
2. Fix a lifetime problem in CSV

### Are these changes tested?

I test it locality. But don't know how to write unittest here. Fell free to help.

### Are there any user-facing changes?

Bugfix

* Closes: #38438

### CODE REVIEW

**Comment by mapleFU** [comment]
@pitrou @bkietz I've tried to fix https://github.com/apache/arrow/issues/38438 here

Also cc @austin3dickey . I've reproduce this bug in my local-machine with ASAN, and fix it with this changing. I'm not so familiar with building Python, you can try to build an verify it here.

File: cpp/src/arrow/csv/reader.cc [hunk 1]
```diff
@@ -1112,15 +1112,16 @@ class AsyncThreadedTableReader
  protected:
   Future<std::shared_ptr<Buffer>> ProcessFirstBuffer() {
     // First block
+    auto self = shared_from_this();
     auto first_buffer_future = buffer_generator_();
-    return first_buffer_future.Then([this](const std::shared_ptr<Buffer>& first_buffer)
+    return first_buffer_future.Then([self](const std::shared_ptr<Buffer>& first_buffer)
```
Comment (bkietz) [inline cpp/src/arrow/csv/reader.cc:1117]:
```suggestion
    return first_buffer_future.Then([self = shared_from_this()](const std::shared_ptr<Buffer>& first_buffer)
```

File: cpp/src/parquet/arrow/reader.cc [hunk 2]
```diff
@@ -1122,11 +1122,12 @@ class RowGroupGenerator {
     } else {
       auto ready = reader->parquet_reader()->WhenBuffered({row_group}, column_indices);
       if (cpu_executor_) ready = cpu_executor_->TransferAlways(ready);
+      auto cpu_executor = cpu_executor_;
       row_group_read =
-          ready.Then([this, reader, row_group,
+          ready.Then([cpu_executor, reader, row_group,
```
Comment (bkietz) [inline cpp/src/parquet/arrow/reader.cc:1127]:
```suggestion
          ready.Then([executor = cpu_executor_, reader, row_group,
```

**Review by bkietz — CHANGES_REQUESTED** [review]
I'll give this a more thorough review tomorrow, for now:

**Comment by austin3dickey** [comment]
@ursabot please benchmark name=dataset-serialize lang=Python

**Review by bkietz — CHANGES_REQUESTED** [review]
Could you please add your reproducer https://github.com/apache/arrow/issues/38438#issuecomment-1779803994 as a test?

**Comment by mapleFU** [comment]
> Could you please add your reproducer https://github.com/apache/arrow/issues/38438#issuecomment-1779803994 as a test?

My code is far from the point  we meet the segment fault. I build a Scanner and call Head to it multiple times.

Let me use Parquet to re-produce it.

**Comment by mapleFU** [comment]
@bkietz How can I mock a stream which `Read` returns a `Future`? `BufferReader` will read instantly. Should I create a mock-reader?

**Comment by bkietz** [comment]
> Should I create a mock-reader?

Yes, this pattern is used elsewhere in testing too. `DelayedBufferReader` should work here, I think
https://github.com/bkietz/arrow/blob/27981976f7dfd4c4ab7f60677f849439e038fc55/cpp/src/arrow/filesystem/mockfs.cc#L243-L254

> And when IO complete, the io thread call the callback, which will trigger the segment fault.

Just adding the regression test to ensure future refactorings don't reintroduce the segfault should be sufficient

**Comment by austin3dickey** [comment]
The [benchmark passed](https://buildkite.com/apache-arrow/arrow-bci-benchmark-on-ursa-i9-9960x/builds/3696#018b6c5a-ba15-4955-b28a-ce4da01a8855/36-6981), which is exactly what we're looking for. Thanks! (Looking into why the notification wasn't posted here)

**Comment by mapleFU** [comment]
> Just adding the regression test to ensure future refactorings don't reintroduce the segfault should be sufficient

Aha the regression test using a large local file

**Comment by mapleFU** [comment]
Do we have some way to break the stack 😅 I've re-produce the case but asan didn't report error.

```
{
  Generator();
}

// do-io task
```

When doing IO task, `ReadRowGroup` doesn't report ASAN error here..

**Comment by mapleFU** [comment]
```c++
class DelayedBufferReader : public ::arrow::io::BufferReader {
 public:
  explicit DelayedBufferReader(const std::shared_ptr<::arrow::Buffer>& buffer)
      : ::arrow::io::BufferReader(buffer) {}

  ::arrow::Future<std::shared_ptr<Buffer>> ReadAsync(const ::arrow::io::IOContext& io_context, int64_t position,
                                            int64_t nbytes) override {
    read_async_count.fetch_add(1);
    std::this_thread::sleep_for(std::chrono::seconds(10));
    return ::arrow::io::BufferReader::ReadAsync(io_context, position, nbytes);
  }

  std::atomic<int> read_async_count{0};
};

TEST(TestArrowReadWrite, ScanContentsGracefulShutdown) {
  ArrowReaderProperties properties = default_arrow_reader_properties();
  properties.set_batch_size(256);
  properties.set_pre_buffer(true);
  properties.set_use_threads(true);
  auto cache_options = ::arrow::io::CacheOptions::LazyDefaults();
  cache_options.hole_size_limit = 1;
  cache_options.range_size_limit = 1;
  properties.set_cache_options(cache_options);
  const int num_rows = 1024;
  const int row_group_size = 16;
  const int num_columns = 1;

  std::shared_ptr<Table> table;
  ASSERT_NO_FATAL_FAILURE(MakeDoubleTable(num_columns, num_rows, 1, &table));

  std::shared_ptr<Buffer> buffer;
  ASSERT_NO_FATAL_FAILURE(WriteTableToBuffer(table, row_group_size,
                                             default_arrow_writer_properties(), &buffer));

  auto mock_input_stream = std::make_shared<DelayedBufferReader>(buffer);
  std::vector<std::unique_ptr<::arrow::DelayedExecutor>> delayed_executors;
  delayed_executors.resize(3);
  // vector of futures
  std::vector<::arrow::Future<std::shared_ptr<::arrow::RecordBatch>>> futures;
  for (int idx = 0; idx < 3; ++idx) {
    delayed_executors[idx] = std::make_unique<::arrow::DelayedExecutor>();
    auto& delayed_executor = *delayed_executors[idx];
    std::shared_ptr<FileReader> reader;
    {
      std::unique_ptr<FileReader> unique_reader;
      FileReaderBuilder builder;
      ASSERT_OK(builder.Open(mock_input_stream));
      ASSERT_OK(builder.properties(properties)->Build(&unique_reader));
      reader = std::move(unique_reader);
    }
    {
      ASSERT_OK_AND_ASSIGN(
          auto batch_generator,
          reader->GetRecordBatchGenerator(reader, {0, 1}, {0}, &delayed_executor, 256));
      auto fut1 = batch_generator();
      auto fut2 = batch_generator();
      futures.push_back(fut1);
      futures.push_back(fut2);
    }
    // clear reader.
    reader = nullptr;
  }
  auto pool = ::arrow::internal::ThreadPool::Make(2).ValueOrDie();
  pool->Submit([&]() {
    for (int idx = 0; idx < 3; ++idx) {
      auto& delayed_executor = *delayed_executors[idx];
      while (!delayed_executor.captured_tasks.empty()) {
        std::cout << "count:" << delayed_executor.captured_tasks.size() << std::endl;
        auto callbacks = std::move(delayed_executor.captured_tasks);
        for (auto& callback : callbacks) {
          std::move(callback)();
        }
      }
    }
  }).ValueOrDie().Wait();
}
```

I've tried an ugly test like this, but though ` reader->GetRecordBatchGenerator` has dtor, the `this->executor_` didn't trigger the ASAN / UBSAN error. Maybe I need to take another ways later.

File: cpp/src/parquet/arrow/reader.cc [hunk 3]
```diff
@@ -1182,7 +1182,7 @@ FileReaderImpl::GetRecordBatchGenerator(std::shared_ptr<FileReader> reader,
                                         int64_t rows_to_readahead) {
   RETURN_NOT_OK(BoundsCheck(row_group_indices, column_indices));
   if (rows_to_readahead < 0) {
-    return Status::Invalid("rows_to_readahead must be > 0");
+    return Status::Invalid("rows_to_readahead must be >= 0");
```
Comment (mapleFU) [inline cpp/src/parquet/arrow/reader.cc:1185]: This can be 0 here.

**Comment by mapleFU** [comment]
Also cannot re-produce by:

```

class DelayedBufferReader : public ::arrow::io::BufferReader {
 public:
  explicit DelayedBufferReader(const std::shared_ptr<::arrow::Buffer>& buffer)
      : ::arrow::io::BufferReader(buffer) {}

  ::arrow::Future<std::shared_ptr<Buffer>> ReadAsync(const ::arrow::io::IOContext& io_context, int64_t position,
                                                     int64_t nbytes) override {
    read_async_count.fetch_add(1);
    std::this_thread::sleep_for(std::chrono::seconds(1));
    return ::arrow::io::BufferReader::ReadAsync(io_context, position, nbytes);
  }

  std::atomic<int> read_async_count{0};
};

TEST_F(TestParquetFileFormat, MultithreadedScanUnsafe) {
  auto reader = GetRecordBatchReader(schema({field("utf8", utf8())}));

  ASSERT_OK_AND_ASSIGN(auto buffer, ParquetFormatHelper::Write(reader.get()));
  auto buffer_reader = std::make_shared<DelayedBufferReader>(buffer);
  auto source = std::make_shared<FileSource>(std::move(buffer_reader), buffer->size());

  auto fragment = MakeFragment(*source);

  auto options = std::make_shared<ScanOptions>();
  auto fragment_scan_options = std::make_shared<ParquetFragmentScanOptions>();
  fragment_scan_options->arrow_reader_properties->set_pre_buffer(true);
  options->fragment_scan_options = fragment_scan_options;
  options->use_threads = true;
  ScannerBuilder builder(ArithmeticDatasetFixture::schema(), fragment, options);

  ASSERT_OK(builder.UseThreads(true));
  ASSERT_OK(builder.BatchSize(128));
  ASSERT_OK_AND_ASSIGN(auto scanner, builder.Finish());

  ASSERT_OK_AND_ASSIGN(auto batch, scanner->Head(8000));
  ASSERT_OK_AND_ASSIGN(batch, scanner->Head(8000 * 10));
  ASSERT_OK_AND_ASSIGN(batch, scanner->Head(8000 * 10));
}
```

**Comment by mapleFU** [comment]
@bkietz I've tried but cannot reproduce it, but currently failed to reproduce without the sample file...

**Comment by mapleFU** [comment]
Sadly I've trigger another problem when trying to re-produce the bug. 😅

```
class DelayedBufferReader : public ::arrow::io::BufferReader {
 public:
  explicit DelayedBufferReader(const std::shared_ptr<::arrow::Buffer>& buffer)
      : ::arrow::io::BufferReader(buffer) {}

  ::arrow::Future<std::shared_ptr<Buffer>> ReadAsync(const ::arrow::io::IOContext& io_context, int64_t position,
                                                     int64_t nbytes) override {
    read_async_count.fetch_add(1);
    auto self = std::dynamic_pointer_cast<DelayedBufferReader>(shared_from_this());
    return DeferNotOk(::arrow::io::internal::SubmitIO(
        io_context, [self, position, nbytes, io_context] {
          std::this_thread::sleep_for(std::chrono::seconds(1));
          return self->::arrow::io::BufferReader::ReadAsync(io_context, position, nbytes);
        }));
  }

  std::atomic<int> read_async_count{0};
};

TEST_F(TestParquetFileFormat, MultithreadedScanUnsafe) {
  auto reader = MakeGeneratedRecordBatch(schema({field("utf8", utf8())}), 10000, 100);

  ASSERT_OK_AND_ASSIGN(auto buffer, ParquetFormatHelper::Write(reader.get()));

  for (int idx = 0; idx < 3; ++idx) {
    auto buffer_reader = std::make_shared<DelayedBufferReader>(buffer);
    auto source = std::make_shared<FileSource>(buffer_reader, buffer->size());
    auto fragment = MakeFragment(*source);
    std::shared_ptr<Scanner> scanner;

    {
      auto options = std::make_shared<ScanOptions>();
      auto fragment_scan_options = std::make_shared<ParquetFragmentScanOptions>();
      fragment_scan_options->arrow_reader_properties->set_pre_buffer(true);
      options->fragment_scan_options = fragment_scan_options;
      options->use_threads = true;
      ScannerBuilder builder(ArithmeticDatasetFixture::schema(), fragment, options);

      ASSERT_OK(builder.UseThreads(true));
      ASSERT_OK(builder.BatchSize(128));
      ASSERT_OK_AND_ASSIGN(scanner, builder.Finish());
    }

    ASSERT_OK_AND_ASSIGN(auto batch, scanner->Head(10000));
    ASSERT_OK_AND_ASSIGN(batch, scanner->Head(80000 * 10));
    //  ASSERT_OK_AND_ASSIGN(batch, scanner->Head(8000 * 10));
    // Random ReadAsync calls
    for (int yy = 0; yy < 100; yy++) {
      [[maybe_unused]] auto ff = buffer_reader->ReadAsync(::arrow::io::IOContext(), 0, 1001);
    }
    scanner = nullptr;
  }
}
```

@bkietz When we have too many Future, the done might notify too many "done", which might cause stack overflow...

**Comment by mapleFU** [comment]
```c++
class DelayedBufferReader : public ::arrow::io::BufferReader {
 public:
  explicit DelayedBufferReader(const std::shared_ptr<::arrow::Buffer>& buffer)
      : ::arrow::io::BufferReader(buffer) {}

  ::arrow::Future<std::shared_ptr<Buffer>> ReadAsync(const ::arrow::io::IOContext& io_context, int64_t position,
                                                     int64_t nbytes) override {
    read_async_count.fetch_add(1);
    auto self = std::dynamic_pointer_cast<DelayedBufferReader>(shared_from_this());
    return DeferNotOk(::arrow::io::internal::SubmitIO(
        io_context, [self, position, nbytes]() -> Result<std::shared_ptr<Buffer>> {
          std::this_thread::sleep_for(std::chrono::seconds(1));
          return self->DoReadAt(position, nbytes);
        }));
  }

  std::atomic<int> read_async_count{0};
};

TEST_F(TestParquetFileFormat, MultithreadedScanUnsafe) {
  auto reader = MakeGeneratedRecordBatch(schema({field("utf8", utf8())}), 10000, 100);

  ASSERT_OK_AND_ASSIGN(auto buffer, ParquetFormatHelper::Write(reader.get()));

  std::vector<Future<>> completes;
  std::vector<std::shared_ptr<arrow::internal::ThreadPool>> pools;

  for (int idx = 0; idx < 3; ++idx) {
    auto buffer_reader = std::make_shared<DelayedBufferReader>(buffer);
    auto source = std::make_shared<FileSource>(buffer_reader, buffer->size());
    auto fragment = MakeFragment(*source);
    std::shared_ptr<Scanner> scanner;

    {
      auto options = std::make_shared<ScanOptions>();
      ASSERT_OK_AND_ASSIGN(auto thread_pool, arrow::internal::ThreadPool::Make(1));
      pools.emplace_back(thread_pool);
      options->io_context = ::arrow::io::IOContext(::arrow::default_memory_pool(), pools.back().get());
      auto fragment_scan_options = std::make_shared<ParquetFragmentScanOptions>();
      fragment_scan_options->arrow_reader_properties->set_pre_buffer(true);

      options->fragment_scan_options = fragment_scan_options;
      options->use_threads = true;
      ScannerBuilder builder(ArithmeticDatasetFixture::schema(), fragment, options);

      ASSERT_OK(builder.UseThreads(true));
      ASSERT_OK(builder.BatchSize(10000));
      ASSERT_OK_AND_ASSIGN(scanner, builder.Finish());
    }

    ASSERT_OK_AND_ASSIGN(auto batch, scanner->Head(10000));
    auto fut = scanner->ScanBatchesUnorderedAsync();
    //  ASSERT_OK_AND_ASSIGN(batch, scanner->Head(8000 * 10));
    // Random ReadAsync calls
    for (int yy = 0; yy < 16; yy++) {
      completes.emplace_back(buffer_reader->ReadAsync(::arrow::io::IOContext(), 0, 1001));
    }
    scanner = nullptr;
  }

  for (auto& f: completes) {
    f.Wait();
  }
}
```

Finally I re-produce the issue successfully with same stack I meet, should I use this @bkietz ? (this test might runs for a long time...)

File: cpp/src/arrow/dataset/file_parquet_test.cc [hunk 4]
```diff
@@ -834,5 +834,72 @@ TEST(TestParquetStatistics, NullMax) {
...
+  for (int idx = 0; idx < 2; ++idx) {
+    auto buffer_reader = std::make_shared<DelayedBufferReader>(buffer);
+    auto source = std::make_shared<FileSource>(buffer_reader, buffer->size());
+    auto fragment = MakeFragment(*source);
+    std::shared_ptr<Scanner> scanner;
+
+    {
+      auto options = std::make_shared<ScanOptions>();
+      ASSERT_OK_AND_ASSIGN(auto thread_pool, arrow::internal::ThreadPool::Make(1));
+      pools.emplace_back(thread_pool);
```
Comment (mapleFU) [inline cpp/src/arrow/dataset/file_parquet_test.cc:874]: pool is for case mentioned in the description

**Comment by mapleFU** [comment]
This is ready for review now :-|

**Review by bkietz — APPROVED** [review]
Thanks, @mapleFU !

**Pull request merged**